### ISSUE TITLE
Consider dropping gfxOS2Font and related classes

### ISSUE THREAD

**Issue by dmik** [bitwiseworks/mozilla-os2#65]
The `gfxOS2Font` and `gfxOS2FontGroup` classes use Freetype to implement glyph rendering on OS/2. However, there is already `gfxFT2Font` and friends (used on other platforms) that also use Freetype. Perhaps we can reuse these classes on OS/2 as well. Since these classes are upstream, they get instant updates and this way will not have to sync our specific OS/2 versions (which are pretty much outdated after updating to 17esr) with any future updates.

For examples of going out of sync see e.g. #21.

**Comment by dmik** [comment]
Note that once this issue's task is done, 4f06effa0171e7b24e0baa17caf7db21aa4bc4da needs to be reverted. See #74 for details.

**Comment by dmik** [comment]
The experiment of trying to switch from `gfxOS2Fonts` to `gfxPangoFonts` within #115 succeeded. We now have a build of Firefox 38 that doesn't crash and doesn't use `gfxOS2Fonts`. It has some problems with charsets still (see https://github.com/bitwiseworks/mozilla-os2/issues/115#issuecomment-159096680) and I will address them in this ticket.

But first I will commit the `gfxOS2Fonts` to `gfxPangoFonts` switch that made FF 38 run (tomorrow, it needs some cleanup).

**Comment by dryeo** [comment]
So far the new fntcfg2 breaks warpsans to workplace sans mapping it seems. This is most visible in the menus which now seem to use Helvetia much like if workplace sans is unavailable. In particular Thunderbird, which I don't think I ever changed the fonts on, looks terrible with all the fonts being changed. SeaMonkey and Firefox aren't much better, especially the menus and tabs.
When the switch was made to Cairo, Peter put some work into maintaining the menus look, mostly by using Alex's workplace sans so that the Mozilla apps fit into the OS/2 look and feel.
Hmm, attempt to upload screenshots fails but if you compare screenshots side by side you'll see that not only do the menus look terrible but also the Google in the search bar and using issue #115 for comparison, the empty glyph boxes also look fuzzy compared to the original fntcfg. Glyph spacing is also different.

**Comment by dmik** [comment]
It no longer does. See http://trac.netlabs.org/ports/ticket/76 for details. But thanks for reporting.

**Comment by dmik** [comment]
The switch to pango is committed in 9b85e3b45ca4052c392c0cfc75ae1bdbf5583015.

**Comment by dryeo** [comment]
On 11/26/15 09:13 AM, Dmitriy Kuminov wrote:
> The switch to pango is committed in 9b85e3b
> <https://github.com/bitwiseworks/mozilla-os2/commit/9b85e3b45ca4052c392c0cfc75ae1bdbf5583015>.

Will you be releasing your pango build? Here autoconf never produced a 
working configure.

**Comment by dmik** [comment]
Sure, pango, cairo and pixman are to be released as RPMs. I have fixed all problems locally.

**Comment by dmik** [comment]
It seems that `FcPatternGetCharSet (FC_CHARSET)` always returns 0 for matched fonts (patterns) in our fontconfig. As a result, pango thinks the font doesn't have any chars and hence the boxes (even for ASCII). I have a feeling that our fontconfig doesn't properly extract charset information from font files (perhaps, because the respective parts are simply missing). Need to dig deeper in there.

**Comment by dmik** [comment]
It turns out that our fontconfig doesn't read charset information from fonts at all (because the whole font enumeration part is different). In particular, our code never calls `FcFreeTypeQuery` used in stock fontconfig to read out the charset information. If I dirty-hack Mozilla to think that every font has every char, it starts working as expected:

![2015-11-27 23 42 18](https://cloud.githubusercontent.com/assets/900726/11447985/c5692dbc-9560-11e5-88f0-975d0633b27b.png)

Now I tend to taking stock fontconfig sources as a whole and only change the small part where they enumerate fonts (and use the PM INI file instead of XML there, as in the current version), leaving all the remaining original logic intact. This will bring us charsets among other things.

Note that we don't get web fonts working out of the box though (see the github page on the screenshot). Maybe its support needs some actual charset information or there is some failure in the font download code. I will check that after I sort out the charset question.

**Comment by dmik** [comment]
Yes, using `FcFreeTypeQuery` manually (which in particular sets up the charset/language information) for each pattern returned from `FcFontMatch` solves the charset problem. So we should just switch to stock fontconfig to get charset selection and matching work correctly. Note also that we will get multiple propery value checks when we do that (mentioned here: https://github.com/bitwiseworks/mozilla-os2/issues/115#issuecomment-159456950).

**Comment by dryeo** [comment]
On 11/27/15 12:49 PM, Dmitriy Kuminov wrote:
> Note that we don't get web fonts working out of the box though (see the
> github page on the screenshot). Maybe its support needs some actual
> charset information or there is some failure in the font download code.
> I will check that after I sort out the charset question.

Did you remember to re-enable the preference?

**Comment by dmik** [comment]
Yes, doesn't help. There has to be something else.

**Comment by dmik** [comment]
Just for the record, we decided to move to the full upstream version of fontconfig instead of using the restricted one. The details are here: http://trac.netlabs.org/ports/ticket/77. And here is Firefox 38 running with new fontconfig (notice how nicely it selected a replacement for WarpSans):

<img width="781" alt="2015-12-04 15 15 50" src="https://cloud.githubusercontent.com/assets/900726/11591809/93e6a40e-9aab-11e5-981e-94bc5acab36f.png">

It must be DjeaVu Sans or such (which I have installed). I will add a default substitution for WarpSans with Workplace Sans to the stock fonts.conf of course.
